### ISSUE TITLE
[Bug]: Cascade state merging returns NaN when all input states are empty

### ISSUE THREAD

**Issue by morluto** [flashinfer-ai/flashinfer#4051]
## Observed behavior

FlashInfer represents an empty attention state as zero V with logsumexp equal to negative infinity. Merging two such states currently produces NaN for both the attention output and logsumexp.

This affects `merge_state`, `merge_state_in_place`, and reductions using `state_t::merge`, including `merge_states` and persistent attention reductions.

## Reproduction

The official trace reference reproduces the failure:

```python
import torch
from flashinfer.trace.templates.cascade import merge_state_trace

v_a = torch.zeros((1, 1, 64), dtype=torch.float16)
v_b = torch.zeros_like(v_a)
s_a = torch.full((1, 1), -torch.inf, dtype=torch.float32)
s_b = torch.full_like(s_a, -torch.inf)

v, s = merge_state_trace.reference(v_a, s_a, v_b, s_b)

print(torch.isnan(v).any().item())  # True
print(torch.isnan(s).any().item())  # True
```

The merge formula evaluates:

```text
s_a = s_b = -inf
s_max = max(s_a, s_b) = -inf
exp2(s_a - s_max) = exp2(-inf - -inf) = NaN
```

## Expected behavior

The union of empty attention states should remain empty:

```text
merge((0, -inf), (0, -inf)) = (0, -inf)
```

This matches the existing behavior for zero input index sets.

## Impact and scope

An empty or fully masked partition can contaminate subsequent cascade reductions with NaN. Finite-state and single-empty merges are unaffected.

## Success criteria

- `merge_state` and `merge_state_in_place` preserve the empty state.
- Both `merge_states` launch strategies preserve the empty state.
- `state_t::merge` does not evaluate `-inf - -inf`.
- Existing finite-state behavior remains unchanged.
